helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java

```diff
@@ -279,4 +287,156 @@ public static void setRebalanceScheduler(String resourceName, boolean isDelayedR
+    final ResourceConfig resourceConfig = clusterData.getResourceConfig(resourceName);
+    final Map<String, Integer> statePriorityMap =
+        clusterData.getStateModelDef(currentIdealState.getStateModelDefRef()).getStatePriorityMap();
+    final Set<AssignableReplica> toBeAssignedReplicas = new HashSet<>();
+
+    for (PartitionWithReplicaCount partitionWithReplicaCount : partitions) {
+      String partitionName = partitionWithReplicaCount.getPartition().getPartitionName();
+      // count current active replicas of the partition
+      Map<String, Integer> activeStateReplicaCount = stateInstanceMap.getOrDefault(partitionName, Collections.emptyMap())
+          .entrySet()
+          .stream()
+          .collect(Collectors.toMap(Map.Entry::getKey,
+              e -> (int) e.getValue().stream().filter(liveEnabledInstances::contains).count()));
+      int activeReplicas = activeStateReplicaCount.values().stream().reduce(Integer::sum).orElse(0);
+      int minActiveReplica = partitionWithReplicaCount.getMinActiveReplica();
+      int replicaGapCount = minActiveReplica - activeReplicas;
+      if (replicaGapCount <= 0) {
+        // delayed rebalance overwrites isn't required, early stop and move on to next partition
+        continue;
+      }
+      // follow the state priority order, add additional replicas to close the gap on replica count
+      Map<String, Integer> stateCountMap = clusterData.getStateModelDef(currentIdealState.getStateModelDefRef())
+          .getStateCountMap(minActiveReplica, minActiveReplica);
+      // follow the priority order of states and prepare additional replicas to be assigned
+      for (String state : priorityOrderedStates) {
+        if (replicaGapCount <= 0) {
+          break;
+        }
+        int priority = statePriorityMap.get(state);
+        int curActiveStateCount = activeStateReplicaCount.getOrDefault(state, 0);
+        for (int i = 0; i < stateCountMap.get(state) - curActiveStateCount && replicaGapCount > 0; i++) {
+          toBeAssignedReplicas.add(
+              new AssignableReplica(clusterData.getClusterConfig(), resourceConfig, partitionName, state, priority));
+          replicaGapCount--;
+        }
+      }
+    }
+    LOG.info("Replicas: {} need to be brought up for rebalance overwrite.", toBeAssignedReplicas);
+    return toBeAssignedReplicas;
+  }
+
+  private static List<String> getPriorityOrderedStates(String resourceName, ResourceControllerDataProvider clusterData) {
+    IdealState currentIdealState = clusterData.getIdealState(resourceName);
+    Map<String, Integer> statePriorityMap =
+        clusterData.getStateModelDef(currentIdealState.getStateModelDefRef()).getStatePriorityMap();
+
+    List<String> priorityOrderedStates = new ArrayList<>(statePriorityMap.keySet());
+    priorityOrderedStates.sort(Comparator.comparing(a -> statePriorityMap.getOrDefault(a, Integer.MAX_VALUE)));
+    return priorityOrderedStates;
```

Comment:
public List<String> getStatesPriorityList();

We have this API in state model definition. You dont need to write this.

Reply:
Thanks a lot. Will update